subprojects/publish/src/main/java/org/gradle/api/publish/internal/ModuleMetadataFileGenerator.java

```diff
@@ -178,10 +181,52 @@ private void writeIdentity(ModuleVersionIdentifier coordinates, ImmutableAttribu
             jsonWriter.name("version");
             jsonWriter.value(ownerCoordinates.getVersion());
             writeAttributes(componentData.attributes, jsonWriter);
+            writeCapabilities(component, jsonWriter);
             jsonWriter.endObject();
         }
     }
 
+    private void writeCapabilities(SoftwareComponent component, JsonWriter jsonWriter) throws IOException {
+        if (component instanceof ComponentWithCapabilities) {
+            List<CapabilityDescriptor> capabilities = ((ComponentWithCapabilities) component).getCapabilities();
+            if (!capabilities.isEmpty()) {
+                jsonWriter.name("capabilities");
+                jsonWriter.beginArray();
+                for (CapabilityDescriptor capability : capabilities) {
+                    writeCapability(jsonWriter, capability);
+                }
+                jsonWriter.endArray();
+            }
+        }
+    }
+
+    private void writeCapability(JsonWriter jsonWriter, CapabilityDescriptor capability) throws IOException {
+        jsonWriter.beginObject();
+        jsonWriter.name("name");
+        jsonWriter.value(capability.getName());
+        writeCapabilityProviders(jsonWriter, capability);
+        String prefer = capability.getPrefer();
+        if (prefer != null) {
+            jsonWriter.name("prefer");
+            jsonWriter.value(prefer);
+        }
+        String reason = capability.getReason();
```

Comment:
Does the reason apply to the preference? Or to the capability itself? Seems like these are different things.
I'm not sure a capability requires a 'reason', but it might require a 'description'.
A 'prefer' could definitely warrant a 'reason'.

Reply:
Yes, the reason applies to the preference. The initial implementation made it clear by allowing to set a reason only if you called "setPreference". However, this introduced significant noise in the API (an inner class) just for the sake of type safety. I felt it was too much and removed the inner class, so `because` is available at the same level as `prefer`, but both are bound together. The Javadocs explain it.